### ISSUE TITLE
Dice coefficient as custom objective function

### ISSUE THREAD

**Issue by razn7** [fchollet/keras#2994]
I need to implement [dice coefficient](https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient) as objective function in keras.
Seems to be I can't do it right.
y_true and y_pred in custom objective function are Tensor Variables not a real data like numpy array , that's why we can operate only with backend functions like K.sum, K.dot and others or tensor functions (T.sum, T.dot...)

I tried at first:

``` python
def dice_loss1(y_true,y_pred): 
 yt = T.flatten(y_true)
 yp = T.flatten(y_pred)
 loss=1.0-2.0*T.dot(yt,yp)/(T.sum(yt)+T.sum(yp))
 return loss;

```

But also I need clip and round prediction data and (T.sum(yt)+T.sum(yp)) can be zero, that's why:

``` python
def dice_loss2(y_true,y_pred): 
 yt = K.flatten(y_true)
 yp = K.round(K.clip(K.flatten(y_pred),0.7,1.0))
 sum=K.sum(yt)+K.sum(yp)
 if (sum): loss=1.0-2.0*K.dot(yt,yp)/sum
 else: loss=0.0
 return loss;

```

I tried to use K.variable and not to use it, nothing new: dice coefficient doesn't work, it calculated wrong. 

Please say where am I wrong? I am not fully understand how to work with backend functions and Tensor Variables.

**Comment by razn7** [comment]
Some my mistakes i corrected and did in a different way . Now code looks like:
```python
def dice_loss3(y_true,y_pred):
 y_true = y_true.astype('int8')
 y_pred=(T.ceil(T.clip(y_pred,0.7,1.0))).astype('int8')
 insec=T.sum(y_true&y_pred)
 sum=T.sum(y_true)+T.sum(y_pred)
 if (sum):
  return (1.0-2.0*insec/sum)
 else:
  return 0.0
```

But what our objective function must return according to using optimizer?
Seems to be values from [0..1]  returns by dice_loss not good for optimizers.

**Comment by razn7** [comment]
I see in the main page in "Support" section: _you can also post bug reports and **feature requests** in **Github issues**_ That's why I post here.

Maybe somebody know how to implement correct dice_loss function with some optimizer?

**Comment by ducha-aiki** [comment]
@razn7 https://github.com/jocicmarko/ultrasound-nerve-segmentation/blob/master/train.py#L19

**Comment by razn7** [comment]
@ducha-aiki already have seen, thanks for answer.
Marko use the same as dice_loss1 with some small differences. He decided the problem where sum1+sum2=0 with smooth parameter but not the main: how to use ceil and clip and make dice work more precisely!? Or this impossible due to[from keras google group]: _the problem is that loss function must be differentiable, and neither round nor ceil are_

**Comment by lemuriandezapada** [comment]
You can't use if statements in the loss function since they're not cuda-able
You have K.switch to emulate tensor-wide if statements

**Comment by pratikgorji** [comment]
what is y_true and y_pred...
in this code , how is he reading y_true and y_pred . Where is he defining y_true, y_pred ? I dint find any. 
https://github.com/jocicmarko/ultrasound-nerve-segmentation/blob/master/train.py#L19

Can anyone say , where he is defining y_true and y_pred ? !!

**Comment by patricio-astudillo** [comment]
@pratikgorji please refer to some theoretical background on what you are trying to do. I might be wrong but your question shows that you don't fully understand the bigger picture here. This is not a bad thing, but you might want to read more about what you are trying to do before writing any code. I would recommend the following link http://cs231n.github.io but there are many other sources available.  Good luck.
